### PR TITLE
feat(contracts): update L1GasPriceOracle for Curie fork

### PR DESCRIPTION
**Pull request by zimpha**
### Purpose or design rationale of this PR

This PR update the `L1GasPriceOracle` for Curie fork.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] No, this PR doesn't involve a new deployment, git tag, docker image tag
- [ ] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: contracts/src/L2/predeploys/L1GasPriceOracle.sol [hunk 1]
```diff
@@ -48,6 +86,27 @@ contract L1GasPriceOracle is OwnableBase, IL1GasPriceOracle {
     /// @notice The address of whitelist contract.
     IWhitelist public whitelist;
 
+    /// @inheritdoc IL1GasPriceOracle
+    uint256 public override l1BlobBaseFee;
+
+    /// @inheritdoc IL1GasPriceOracle
+    uint256 public override commitScalar;
```
Comment (Thegaram) [inline contracts/src/L2/predeploys/L1GasPriceOracle.sol:93]: What's the consideration for not reusing `scalar` as `commitScalar`? (Either way works, just curious.)

Comment (zimpha) [reply contracts/src/L2/predeploys/L1GasPriceOracle.sol:93]: I don't want to mix the meaning of `scalar` before and after Curie fork.

File: contracts/src/L2/predeploys/L1GasPriceOracle.sol [hunk 2]
```diff
@@ -17,6 +17,28 @@ contract L1GasPriceOracle is OwnableBase, IL1GasPriceOracle {
...
+    /// @dev Thrown when the l1 fee overhead exceed `MAX_OVERHEAD`.
+    error ErrExceedMaxOverhead();
+
+    /// @dev Thrown when the l1 fee scalar exceed `MAX_SCALAR`.
+    error ErrExceedMaxScalar();
+
+    /// @dev Thrown when the caller is not whitelisted.
+    error ErrCallerNotWhitelisted();
+
+    /// @dev Thrown when we enable Curie fork in Curie fork.
```
Comment (colinlyguo) [inline contracts/src/L2/predeploys/L1GasPriceOracle.sol]:
```suggestion
    /// @dev Thrown when we enable Curie fork after Curie fork.
```

File: contracts/src/L2/predeploys/L1GasPriceOracle.sol [hunk 3]
```diff
@@ -48,6 +86,27 @@ contract L1GasPriceOracle is OwnableBase, IL1GasPriceOracle {
...
+    uint256 public override l1BlobBaseFee;
+
+    /// @inheritdoc IL1GasPriceOracle
+    uint256 public override commitScalar;
+
+    /// @inheritdoc IL1GasPriceOracle
+    uint256 public override blobScalar;
+
+    /// @notice Indicates whether the network has gone through the Curie upgrade.
+    bool public isCurie;
```
Comment (colinlyguo) [inline contracts/src/L2/predeploys/L1GasPriceOracle.sol:99]: So need to set this slot to `true` in the hardfork block?

**Review by Thegaram — DISMISSED** [review]

File: contracts/src/L2/predeploys/L1GasPriceOracle.sol (hunk 3 above)
Comment (Thegaram) [reply contracts/src/L2/predeploys/L1GasPriceOracle.sol:99]: Set here: https://github.com/scroll-tech/go-ethereum/pull/755/files#diff-bff5ba79b7b5a8983f6156c393a80a9f3283f7692ff6f641957b5e8b5752d46aR19

File: contracts/src/L2/predeploys/L1GasPriceOracle.sol [hunk 4]
```diff
@@ -90,12 +143,22 @@ contract L1GasPriceOracle is OwnableBase, IL1GasPriceOracle {
...
+    /// @inheritdoc IL1GasPriceOracle
+    function setL1BaseFeeAndBlobBaseFee(
+        uint256 _l1BaseFee,
+        uint256 _l1BlobBaseFee
+    ) external override onlyWhitelistedSender {
         l1BaseFee = _l1BaseFee;
+        l1BlobBaseFee = _l1BlobBaseFee;
 
         emit L1BaseFeeUpdated(_l1BaseFee);
+        emit L1BlobBaseFeeUpdated(_l1BlobBaseFee);
```
Comment (Thegaram) [inline contracts/src/L2/predeploys/L1GasPriceOracle.sol:161]: This event will be missing for the initial setting, not sure if that's a concern. Same for `CommitScalarUpdated` and `BlobScalarUpdated`.

**Review by Thegaram — APPROVED** [review]

**Review by icemelon — APPROVED** [review]

File: contracts/src/L2/predeploys/L1GasPriceOracle.sol (hunk 4 above)
Comment (zimpha) [reply contracts/src/L2/predeploys/L1GasPriceOracle.sol:161]: Not a big problem to me. Those who want to index evnets from this contract should also be aware of the hardfork block. They can set initial values at the hardfork block.

**Review by colinlyguo — APPROVED** [review]

File: contracts/src/L2/predeploys/L1GasPriceOracle.sol [hunk 5]
```diff
@@ -129,4 +210,57 @@ contract L1GasPriceOracle is OwnableBase, IL1GasPriceOracle {
...
+        return (_l1Fee * scalar) / PRECISION;
+    }
+
+    /// @dev Internal function to compute the L1 portion of the fee based on the size of the rlp encoded input
+    ///   transaction, the current L1 base fee, and the various dynamic parameters, after Curie fork.
+    /// @param _data Signed fully RLP-encoded transaction to get the L1 fee for.
+    /// @return L1 fee that should be paid for the tx
+    function _getL1FeeCurie(bytes memory _data) private view returns (uint256) {
+        // We have bounded the value of `commitScalar` and `blobScalar`, the whole expression won't overflow.
+        return (commitScalar * l1BaseFee + blobScalar * _data.length * l1BlobBaseFee) / PRECISION;
```
Comment (DreamWuGit) [inline contracts/src/L2/predeploys/L1GasPriceOracle.sol:264]: this `_data.length` refer to tx call data length? the tx calls this contract method

Comment (zimpha) [reply contracts/src/L2/predeploys/L1GasPriceOracle.sol:264]: Signed fully RLP-encoded transaction to get the L1 fee for.

**Pull request merged**